File: solutions/kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.go
// @Title: Kth Smallest Element in a BST
// @Author: colinjxc
// @Date: 2022-01-30T20:44:43+08:00
// @URL: https://leetcode-cn.com/problems/kth-smallest-element-in-a-bst


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int {
    stack := make([]*TreeNode, 0)
    for root != nil || len(stack) > 0 {
        for root != nil {
            stack = append(stack, root)
            root = root.Left
        }
        root = stack[len(stack)-1]
        stack = stack[0:len(stack)-1]
        k--
        if k == 0 {
            return root.Val
        }
        root = root.Right
    }
    return 0
}

File: solutions/kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.cpp
// @Title: Kth Smallest Element in a BST
// @Author: colinjxc
// @Date: 2022-01-30T20:26:32+08:00
// @URL: https://leetcode-cn.com/problems/kth-smallest-element-in-a-bst


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        std::vector<TreeNode*> stack;
        while (root || !stack.empty()) {
            while (root) {
                stack.push_back(root);
                root = root->left;
            }
            root = stack.back();
            stack.pop_back();
            if (--k == 0) {
                return root->val;
            }
            root = root->right;
        }
        return -1;
    }
};
